\documentclass[letterpaper]{article}
\usepackage{booktabs}
\setlength{\abovecaptionskip}{6pt}
\setlength{\belowcaptionskip}{6pt}
% Different font in captions
\newcommand{\captionfonts}{\footnotesize}
\makeatletter  % Allow the use of @ in command names
\long\def\@makecaption#1#2{%
  \vskip\abovecaptionskip
  \sbox\@tempboxa{{\captionfonts #1: #2}}%
  \ifdim \wd\@tempboxa >\hsize
    {\captionfonts #1: #2\par}
  \else
    \hbox to\hsize{\hfil\box\@tempboxa\hfil}%
  \fi
  \vskip\belowcaptionskip}
\makeatother   % Cancel the effect of \makeatletter
\usepackage{multirow}
\begin{document}
\begin{table}[tp]
\scriptsize
\begin{tabular*}{1.00\textwidth}{@{\extracolsep{\fill}} lrrrrr}
\multicolumn{6}{c}{\small{Mapping time: workstation}} \\[3pt]
\toprule
 & \multirow{2}{*}{CPU Time} & Wall clock & Bowtie  & Peak virtual & Reads  \\
 &                            & time       & Speedup & memory usage & mapped \\[3pt]
\toprule
Bowtie & 16m:55s & 19m:13s & - & 1,169 MB & 74.3\%\\\midrule
Maq & 11h:15m:58s & 11h:22m:02s & 35.5x & 804 MB & 78.0\%\\
\toprule\\[1pt]
\multicolumn{6}{c}{\small{Mapping time: server}} \\[3pt]
\toprule
 & \multirow{2}{*}{CPU Time} & Wall clock & Bowtie  & Peak virtual & Reads  \\
 &                            & time       & Speedup & memory usage & mapped \\[3pt]
\toprule
Bowtie & 18m:05s & 20m:02s & - & 1,169 MB & 74.3\%\\\midrule
Maq & 18h:49m:07s & 18h:50m:16s & 56.4x & 804 MB & 78.0\%\\
\bottomrule\\[1pt]
\end{tabular*}
\begin{tabular*}{1.00\textwidth}{@{\extracolsep{\fill}} rrrr}
\multicolumn{4}{c}{\small{Reads mapped}} \\[3pt]
\toprule 
By Maq or      & By both Maq & By Bowtie   & By Maq         \\ 
Bowtie or both & and Bowtie  & but not Maq & but not Bowtie \\ 
\toprule
6,553,324 & 6,240,692 (95.2\%) & 1,435 (0.0\%) & 311,197 (4.7\%) \\ 
\bottomrule
\end{tabular*}
\caption{Performance and sensitivity measurements for mapping an unfiltered
set of 8.96M 35bp Illumina/Solexa reads
against the whole human genome on a single CPU of a workstation with a 2.4 GHz
Intel Core 2 Q6600 processor and 2 GB of RAM, and on a server with a 2.4 GHz
AMD Opteron 850 processor and 32 GB of RAM. Bowtie speedup is calculated with
respect to wall clock time. Both CPU time and wall clock times are included to
demonstrate that no one tool suffers disproportionately from I/O pauses or
contention with other processes. Maq indexes its inputs as it maps,
whereas Bowtie is provided with a pre-built index of the
genome.  The cost of building the Bowtie index is not included in these timings
since we expect that in practice that cost will be rapidly amortized across
multiple mapping jobs. Reads are taken from the 1000-Genomes project pilot via
the NCBI Short Read archive, accession \#SRR001115 and trimmed to 35bps.
Reference sequences were the contigs of Genbank human genome build 36.3. For
the Maq runs, the reads were first
divided into chunks of 2M reads each, as per the Maq Manual. Maq
v0.6.6 was used. }
\end{table}
\end{document}
