\documentclass[letterpaper]{article}
\usepackage{booktabs}
\setlength{\abovecaptionskip}{6pt}
\setlength{\belowcaptionskip}{6pt}
% Different font in captions
\newcommand{\captionfonts}{\footnotesize}
\makeatletter  % Allow the use of @ in command names
\long\def\@makecaption#1#2{%
  \vskip\abovecaptionskip
  \sbox\@tempboxa{{\captionfonts #1: #2}}%
  \ifdim \wd\@tempboxa >\hsize
    {\captionfonts #1: #2\par}
  \else
    \hbox to\hsize{\hfil\box\@tempboxa\hfil}%
  \fi
  \vskip\belowcaptionskip}
\makeatother   % Cancel the effect of \makeatletter
\usepackage{multirow}
\begin{document}
\begin{table}[tp]
\scriptsize
\begin{tabular}{lrrrrr}
\toprule
 & \multirow{2}{*}{CPU Time} & Wall clock & Bowtie  & Peak virtual & Reads  \\
 &                            & time       & Speedup & memory usage & mapped \\[3pt]
\toprule
Bowtie -v 2 & 23m:11s & 24m:02s & - & 1,149 MB & 70.1\%\\\midrule
Soap & 106h:57m:37s & 107h:12m:28s & 267.6x & 13,619 MB & 70.6\%\\
\bottomrule
\end{tabular}
\caption{Performance measurements for mapping 8.96M 35bp Illumina/Solexa reads against the whole human genome on a single CPU of a server with a 2.4 GHz AMD Opteron 850 processor and 32 GB of RAM. Bowtie speedup is calculated with respect to wall clock time. Both CPU time and wall clock times are included to demonstrate that no one tool suffers disproportionately from I/O pauses or contention with other processes on the system. Note that Maq (resp. Soap) indexes the reads (resp. genome) as it maps, whereas the Bowtie mapper requires a pre-built index of the genome.  The cost of building the Bowtie index is not included in these timings since we expect that in practice that cost will be rapidly amortized across multiple mapping jobs, or that the researcher will simply download a pre-built index from a shared repository in much less time than is required to build from scratch. Reads are taken from the 1000-Genomes project pilot via the NCBI Short Read archive, accession \#SRR001115 and trimmed to 35bps. Reference sequences were the contigs of Genbank human genome build 36.3. For the Maq runs, the reads were first divided into chunks of 2M reads each, as per the Maq Manual. Soap v1.10 and Maq v0.6.6 were used. }
\end{table}
\end{document}
