\documentclass[letterpaper]{article}
\usepackage{booktabs}
\setlength{\abovecaptionskip}{6pt}
\setlength{\belowcaptionskip}{6pt}
% Different font in captions
\newcommand{\captionfonts}{\footnotesize}
\makeatletter  % Allow the use of @ in command names
\long\def\@makecaption#1#2{%
  \vskip\abovecaptionskip
  \sbox\@tempboxa{{\captionfonts #1: #2}}%
  \ifdim \wd\@tempboxa >\hsize
    {\captionfonts #1: #2\par}
  \else
    \hbox to\hsize{\hfil\box\@tempboxa\hfil}%
  \fi
  \vskip\belowcaptionskip}
\makeatother   % Cancel the effect of \makeatletter
\usepackage{multirow}
\begin{document}
\begin{table}[tp]
\scriptsize
\begin{tabular}{lrrrrr}
\multicolumn{6}{c}{\small{Mapping time: server}} \\[3pt]
\toprule
 & \multirow{2}{*}{CPU Time} & Wall clock & Bowtie  & Peak virtual & Reads  \\
 &                            & time       & Speedup & memory usage & mapped \\[3pt]
\toprule
Bowtie & 24m:28s & 25m:26s & - & 1,169 MB & 76.5\%\\\midrule
Soap & 106h:57m:37s & 107h:12m:28s & 252.9x & 13,619 MB & 70.6\%\\
\bottomrule\\[1pt]
\end{tabular}
\begin{tabular}{rrrr}
\multicolumn{4}{c}{\small{Reads mapped}} \\[3pt] 
By Soap or     & By both Soap & By Bowtie    & By Soap         \\ 
Bowtie or both & and Bowtie   & but not Soap & but not Bowtie \\ 
\toprule
6,332,570 & 6,268,551 (99.0\%) & 6,957 (0.1\%) & 57,062 (0.9\%) \\ 
\bottomrule
\end{tabular}
\caption{Performance and sensitivity measurements for mapping an unfiltered
set of 8.96M 35bp Illumina/Solexa reads
against the whole human genome on a single CPU of a workstation with a 2.4 GHz
Intel Core 2 Q6600 processor and 2 GB of RAM, and on a server with a 2.4 GHz
AMD Opteron 850 processor and 32 GB of RAM. Bowtie speedup is calculated with
respect to wall clock time. Both CPU time and wall clock times are included to
demonstrate that no one tool suffers disproportionately from I/O pauses or
contention with other processes. Soap indexes its inputs
as it maps, whereas Bowtie must be provided with a pre-built index of the
genome.  The cost of building the Bowtie index is not included in these timings
since we expect that in practice that cost will be rapidly amortized across
multiple mapping jobs. Reads are taken from the 1000-Genomes project pilot via
the NCBI Short Read archive, accession \#SRR001115 and trimmed to 35bps.
Reference sequences were the contigs of Genbank human genome build 36.3. Soap
was not run on the workstation because its memory footprint would have exceeded
the physical RAM of the workstation. Soap v1.10 was used. }
\end{table}
\end{document}
