\documentclass[letterpaper]{article}
\usepackage{booktabs}
\setlength{\abovecaptionskip}{6pt}
\setlength{\belowcaptionskip}{6pt}
% Different font in captions
\newcommand{\captionfonts}{\footnotesize}
\makeatletter  % Allow the use of @ in command names
\long\def\@makecaption#1#2{%
  \vskip\abovecaptionskip
  \sbox\@tempboxa{{\captionfonts #1: #2}}%
  \ifdim \wd\@tempboxa >\hsize
    {\captionfonts #1: #2\par}
  \else
    \hbox to\hsize{\hfil\box\@tempboxa\hfil}%
  \fi
  \vskip\belowcaptionskip}
\makeatother   % Cancel the effect of \makeatletter
\usepackage{multirow}
\begin{document}
\begin{table}[tp]
\scriptsize
\begin{tabular}{lrrrr}\multicolumn{5}{c}{2.4 GHz Intel Core 2 workstation with 2 GB of RAM}\\
\toprule
 & \multirow{2}{*}{CPU Time} & Wall clock & Bowtie  & Reads  \\
 &                            & time       & Speedup & mapped \\[3pt]
\toprule
Bowtie -n 1 & 8m:15s & 10m:05s & - & 71.6\%\\\midrule
Maq -n 1 & 6h:36m:57s & 6h:38m:53s & 39.6x & 70.3\%\\\midrule
Bowtie & 16m:58s & 19m:31s & - & 75.3\%\\\midrule
Maq & 17h:46m:35s & 17h:53m:07s & 55.0x & 73.8\%\\
\bottomrule
\end{tabular}
\caption{CPU time for mapping 8.96M 35bp Illumina/Solexa reads against the whole human genome on a workstation with a 2.40GHz Intel Core 2 Q6600 processor and 2 GB of RAM. Reads were originally extracted as part of the 1000-Genomes project pilot. They were downloaded from the NCBI Short Read archive, accession \#SRR001115. Reference sequences were the contigs of Genbank human genome build 36.3. Soap was not run against the whole-human reference because its memory footprint exceeds physical RAM. For the Maq runs, the reads were first divided into chunks of 2M reads each, as per the Maq Manual.}
\end{table}
\end{document}
